elasticdl/python/common/tensor.py

```diff
@@ -0,0 +1,116 @@
+import numpy as np
+import tensorflow as tf
+
+from elasticdl.proto import elasticdl_pb2
+from elasticdl.python.common.dtypes import (
+    dtype_numpy_to_tensor,
+    dtype_tensor_to_numpy,
+)
+
+
+def serialize_tensor(tensor, tensor_pb):
+    """Convert Tensor to Tensor PB"""
+    dtype = dtype_numpy_to_tensor(tensor.values.dtype)
+    if not dtype:
+        raise ValueError(
+            "Dtype of ndarray %s is not supported", tensor.values.dtype
+        )
+    tensor_pb.dtype = dtype
+    tensor_pb.dim.extend(tensor.values.shape)
+    tensor_pb.content = tensor.values.tobytes()
+    if tensor.is_indexed_slices():
+        tensor_pb.indices.extend(tensor.indices)
+    if tensor.name:
+        tensor_pb.name = tensor.name
+
+
+def deserialize_tensor_pb(tensor_pb, tensor):
+    """Create a Tensor instance from Tensor proto message.
+
+    Note: upon return, the input tensor message is reset and underlying buffer
+    passed to the returned ndarray.
+    """
+
+    if not tensor_pb.dim:
+        raise ValueError("Tensor PB has no dim defined")
+
+    dtype = dtype_tensor_to_numpy(tensor_pb.dtype)
+    # Check that the buffer size agrees with dimensions.
+    size = dtype.itemsize
+    for d in tensor_pb.dim:
+        size *= d
+    if size != len(tensor_pb.content):
+        raise ValueError(
+            "Tensor PB size mismatch, dim: %s, len(content): %d",
+            tensor_pb.dim,
+            len(tensor_pb.content),
+        )
+    tensor = Tensor()
+    tensor.set(
+        np.ndarray(shape=tensor_pb.dim, dtype=dtype, buffer=tensor_pb.content),
+        tensor_pb.indices,
```

Comment:
Here, we directly set `tensor_pb.indices` to `tensor.indices`. So the indices type of tensor is actually a Repeated Proto type. This is not we expected. The indices type of tensor should be a 1-D numpy ndarry.

```python
indices = np.array([i, for i in tensor_pb.indices])
tensor.set(..., indices,...)
```

Reply:
Good point. And it seems that `indices = np.array(tensor_pb.indices)` works.

Reply:
Modified. Also added a unit test about the type of `tensor.incides` and `tensor.values`.